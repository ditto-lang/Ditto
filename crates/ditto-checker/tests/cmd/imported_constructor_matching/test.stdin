module Test exports (..);

import Dep;
import Maybe;


zero = Dep.n_params(Dep.A);

abc_to_bool = fn (abc) -> 
    match abc with
    | Dep.C(bool, _) -> Maybe.Just(bool)
    | _ -> Maybe.Nothing
    end;