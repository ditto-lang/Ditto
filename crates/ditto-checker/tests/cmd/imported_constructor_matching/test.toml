bin.name = "ditto-checker-testbin-check-module"
args = ["imported_constructor_matching.ditto"]
fs.sandbox = true
