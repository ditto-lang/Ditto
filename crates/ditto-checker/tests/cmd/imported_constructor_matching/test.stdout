{
  "module_name": [
    "Test"
  ],
  "exports": {
    "types": {},
    "constructors": {},
    "values": {
      "abc_to_bool": {
        "doc_comments": [],
        "doc_position": 0,
        "value_type": {
          "type": "Function",
          "data": {
            "parameters": [
              {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "Constructor",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            "Type"
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Dep"
                          ]
                        ],
                        "value": "ABC"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "ABC"
                      }
                    }
                  },
                  "arguments": [
                    {
                      "type": "Variable",
                      "data": {
                        "variable_kind": "Type",
                        "var": 1,
                        "source_name": null
                      }
                    }
                  ]
                }
              }
            ],
            "return_type": {
              "type": "Call",
              "data": {
                "function": {
                  "type": "Constructor",
                  "data": {
                    "constructor_kind": {
                      "Function": {
                        "parameters": [
                          "Type"
                        ]
                      }
                    },
                    "canonical_value": {
                      "module_name": [
                        null,
                        [
                          "Maybe"
                        ]
                      ],
                      "value": "Maybe"
                    },
                    "source_value": {
                      "module_name": null,
                      "value": "Maybe"
                    }
                  }
                },
                "arguments": [
                  {
                    "type": "PrimConstructor",
                    "data": "Bool"
                  }
                ]
              }
            }
          }
        }
      },
      "zero": {
        "doc_comments": [],
        "doc_position": 1,
        "value_type": {
          "type": "PrimConstructor",
          "data": "Int"
        }
      }
    }
  },
  "types": {},
  "constructors": {},
  "values": {
    "abc_to_bool": {
      "doc_comments": [],
      "name_span": {
        "start_offset": 84,
        "end_offset": 95
      },
      "expression": {
        "expression": "Function",
        "data": {
          "span": {
            "start_offset": 98,
            "end_offset": 203
          },
          "function_type": {
            "type": "Function",
            "data": {
              "parameters": [
                {
                  "type": "Call",
                  "data": {
                    "function": {
                      "type": "Constructor",
                      "data": {
                        "constructor_kind": {
                          "Function": {
                            "parameters": [
                              "Type"
                            ]
                          }
                        },
                        "canonical_value": {
                          "module_name": [
                            null,
                            [
                              "Dep"
                            ]
                          ],
                          "value": "ABC"
                        },
                        "source_value": {
                          "module_name": null,
                          "value": "ABC"
                        }
                      }
                    },
                    "arguments": [
                      {
                        "type": "Variable",
                        "data": {
                          "variable_kind": "Type",
                          "var": 1,
                          "source_name": null
                        }
                      }
                    ]
                  }
                }
              ],
              "return_type": {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "Constructor",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            "Type"
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Maybe"
                          ]
                        ],
                        "value": "Maybe"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "Maybe"
                      }
                    }
                  },
                  "arguments": [
                    {
                      "type": "PrimConstructor",
                      "data": "Bool"
                    }
                  ]
                }
              }
            }
          },
          "binders": [
            [
              {
                "Variable": {
                  "span": {
                    "start_offset": 102,
                    "end_offset": 105
                  },
                  "name": "abc"
                }
              },
              {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "Constructor",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            "Type"
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Dep"
                          ]
                        ],
                        "value": "ABC"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "ABC"
                      }
                    }
                  },
                  "arguments": [
                    {
                      "type": "Variable",
                      "data": {
                        "variable_kind": "Type",
                        "var": 1,
                        "source_name": null
                      }
                    }
                  ]
                }
              }
            ]
          ],
          "body": {
            "expression": "Match",
            "data": {
              "span": {
                "start_offset": 115,
                "end_offset": 203
              },
              "match_type": {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "Constructor",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            "Type"
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Maybe"
                          ]
                        ],
                        "value": "Maybe"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "Maybe"
                      }
                    }
                  },
                  "arguments": [
                    {
                      "type": "PrimConstructor",
                      "data": "Bool"
                    }
                  ]
                }
              },
              "expression": {
                "expression": "LocalVariable",
                "data": {
                  "span": {
                    "start_offset": 121,
                    "end_offset": 124
                  },
                  "variable_type": {
                    "type": "Call",
                    "data": {
                      "function": {
                        "type": "Constructor",
                        "data": {
                          "constructor_kind": {
                            "Function": {
                              "parameters": [
                                "Type"
                              ]
                            }
                          },
                          "canonical_value": {
                            "module_name": [
                              null,
                              [
                                "Dep"
                              ]
                            ],
                            "value": "ABC"
                          },
                          "source_value": {
                            "module_name": null,
                            "value": "ABC"
                          }
                        }
                      },
                      "arguments": [
                        {
                          "type": "Variable",
                          "data": {
                            "variable_kind": "Type",
                            "var": 1,
                            "source_name": null
                          }
                        }
                      ]
                    }
                  },
                  "variable": "abc"
                }
              },
              "arms": [
                [
                  {
                    "ImportedConstructor": {
                      "span": {
                        "start_offset": 136,
                        "end_offset": 150
                      },
                      "constructor": {
                        "module_name": [
                          null,
                          [
                            "Dep"
                          ]
                        ],
                        "value": "C"
                      },
                      "arguments": [
                        {
                          "Variable": {
                            "span": {
                              "start_offset": 142,
                              "end_offset": 146
                            },
                            "name": "bool"
                          }
                        },
                        {
                          "Unused": {
                            "span": {
                              "start_offset": 148,
                              "end_offset": 149
                            },
                            "unused_name": "_"
                          }
                        }
                      ]
                    }
                  },
                  {
                    "expression": "Call",
                    "data": {
                      "span": {
                        "start_offset": 154,
                        "end_offset": 170
                      },
                      "call_type": {
                        "type": "Call",
                        "data": {
                          "function": {
                            "type": "Constructor",
                            "data": {
                              "constructor_kind": {
                                "Function": {
                                  "parameters": [
                                    "Type"
                                  ]
                                }
                              },
                              "canonical_value": {
                                "module_name": [
                                  null,
                                  [
                                    "Maybe"
                                  ]
                                ],
                                "value": "Maybe"
                              },
                              "source_value": {
                                "module_name": null,
                                "value": "Maybe"
                              }
                            }
                          },
                          "arguments": [
                            {
                              "type": "PrimConstructor",
                              "data": "Bool"
                            }
                          ]
                        }
                      },
                      "function": {
                        "expression": "ImportedConstructor",
                        "data": {
                          "span": {
                            "start_offset": 154,
                            "end_offset": 164
                          },
                          "constructor_type": {
                            "type": "Function",
                            "data": {
                              "parameters": [
                                {
                                  "type": "PrimConstructor",
                                  "data": "Bool"
                                }
                              ],
                              "return_type": {
                                "type": "Call",
                                "data": {
                                  "function": {
                                    "type": "Constructor",
                                    "data": {
                                      "constructor_kind": {
                                        "Function": {
                                          "parameters": [
                                            "Type"
                                          ]
                                        }
                                      },
                                      "canonical_value": {
                                        "module_name": [
                                          null,
                                          [
                                            "Maybe"
                                          ]
                                        ],
                                        "value": "Maybe"
                                      },
                                      "source_value": {
                                        "module_name": null,
                                        "value": "Maybe"
                                      }
                                    }
                                  },
                                  "arguments": [
                                    {
                                      "type": "PrimConstructor",
                                      "data": "Bool"
                                    }
                                  ]
                                }
                              }
                            }
                          },
                          "constructor": {
                            "module_name": [
                              null,
                              [
                                "Maybe"
                              ]
                            ],
                            "value": "Just"
                          }
                        }
                      },
                      "arguments": [
                        {
                          "Expression": {
                            "expression": "LocalVariable",
                            "data": {
                              "span": {
                                "start_offset": 165,
                                "end_offset": 169
                              },
                              "variable_type": {
                                "type": "PrimConstructor",
                                "data": "Bool"
                              },
                              "variable": "bool"
                            }
                          }
                        }
                      ]
                    }
                  }
                ],
                [
                  {
                    "Unused": {
                      "span": {
                        "start_offset": 177,
                        "end_offset": 178
                      },
                      "unused_name": "_"
                    }
                  },
                  {
                    "expression": "ImportedConstructor",
                    "data": {
                      "span": {
                        "start_offset": 182,
                        "end_offset": 195
                      },
                      "constructor_type": {
                        "type": "Call",
                        "data": {
                          "function": {
                            "type": "Constructor",
                            "data": {
                              "constructor_kind": {
                                "Function": {
                                  "parameters": [
                                    "Type"
                                  ]
                                }
                              },
                              "canonical_value": {
                                "module_name": [
                                  null,
                                  [
                                    "Maybe"
                                  ]
                                ],
                                "value": "Maybe"
                              },
                              "source_value": {
                                "module_name": null,
                                "value": "Maybe"
                              }
                            }
                          },
                          "arguments": [
                            {
                              "type": "PrimConstructor",
                              "data": "Bool"
                            }
                          ]
                        }
                      },
                      "constructor": {
                        "module_name": [
                          null,
                          [
                            "Maybe"
                          ]
                        ],
                        "value": "Nothing"
                      }
                    }
                  }
                ]
              ]
            }
          }
        }
      }
    },
    "zero": {
      "doc_comments": [],
      "name_span": {
        "start_offset": 55,
        "end_offset": 59
      },
      "expression": {
        "expression": "Call",
        "data": {
          "span": {
            "start_offset": 62,
            "end_offset": 81
          },
          "call_type": {
            "type": "PrimConstructor",
            "data": "Int"
          },
          "function": {
            "expression": "ImportedVariable",
            "data": {
              "span": {
                "start_offset": 62,
                "end_offset": 74
              },
              "variable_type": {
                "type": "Function",
                "data": {
                  "parameters": [
                    {
                      "type": "Call",
                      "data": {
                        "function": {
                          "type": "Constructor",
                          "data": {
                            "constructor_kind": {
                              "Function": {
                                "parameters": [
                                  "Type"
                                ]
                              }
                            },
                            "canonical_value": {
                              "module_name": [
                                null,
                                [
                                  "Dep"
                                ]
                              ],
                              "value": "ABC"
                            },
                            "source_value": {
                              "module_name": null,
                              "value": "ABC"
                            }
                          }
                        },
                        "arguments": [
                          {
                            "type": "Variable",
                            "data": {
                              "variable_kind": "Type",
                              "var": 5,
                              "source_name": null
                            }
                          }
                        ]
                      }
                    }
                  ],
                  "return_type": {
                    "type": "PrimConstructor",
                    "data": "Int"
                  }
                }
              },
              "variable": {
                "module_name": [
                  null,
                  [
                    "Dep"
                  ]
                ],
                "value": "n_params"
              }
            }
          },
          "arguments": [
            {
              "Expression": {
                "expression": "ImportedConstructor",
                "data": {
                  "span": {
                    "start_offset": 75,
                    "end_offset": 80
                  },
                  "constructor_type": {
                    "type": "Call",
                    "data": {
                      "function": {
                        "type": "Constructor",
                        "data": {
                          "constructor_kind": {
                            "Function": {
                              "parameters": [
                                "Type"
                              ]
                            }
                          },
                          "canonical_value": {
                            "module_name": [
                              null,
                              [
                                "Dep"
                              ]
                            ],
                            "value": "ABC"
                          },
                          "source_value": {
                            "module_name": null,
                            "value": "ABC"
                          }
                        }
                      },
                      "arguments": [
                        {
                          "type": "Variable",
                          "data": {
                            "variable_kind": "Type",
                            "var": 5,
                            "source_name": null
                          }
                        }
                      ]
                    }
                  },
                  "constructor": {
                    "module_name": [
                      null,
                      [
                        "Dep"
                      ]
                    ],
                    "value": "A"
                  }
                }
              }
            }
          ]
        }
      }
    }
  },
  "values_toposort": [
    "abc_to_bool",
    "zero"
  ]
}